build_modules/lib/src/module_builder.dart

```diff
@@ -7,15 +7,95 @@ import 'dart:convert';
 
 import 'package:build/build.dart';
 
+import 'meta_module.dart';
+import 'meta_module_builder.dart';
 import 'modules.dart';
 
 /// The extension for serialized module assets.
 const moduleExtension = '.module';
 
+/// A map of primary [AssetId] to corresponding [Module].
+final _primaryToClean =
+    new Resource<Map<AssetId, Module>>(() => {}, dispose: (map) => map.clear());
+
+/// Set of all [MetaModule]s which have been processed.
+final _readMetas = new Resource<Set<AssetId>>(() => new Set<AssetId>(),
+    dispose: (s) => s.clear());
+
+/// Map of [AssetId] to corresponding primary [AssetId] within the same
+/// [Module].
+final _assetToPrimary = new Resource<Map<AssetId, AssetId>>(() => {},
+    dispose: (map) => map.clear());
+
+void _cacheAssetToPrimary(
+    MetaModule meta, Map<AssetId, AssetId> assetToPrimary) {
+  for (var module in meta.modules) {
+    for (var source in module.sources) {
+      assetToPrimary[source] = module.primarySource;
+    }
+  }
+}
+
+/// Removes dependencies from the provided [Module] so that only those with
+/// corresponding module files remain.
```

Comment:
nit: Note that this doesn't just remove dependencies - it modifies them to point at the primary source

Reply:
Should be more clear now.